designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,469 @@
+
+// Verify files
+const results = await linter.lintFiles(patterns)
+// Format and write the results
+process.stdout.write(formatter.format(results))
+```
+
+Currently, the wrapper does:
+
+1. sort given lint results.
+1. create `rulesMeta`.
+
+```js
+class LinterShell {
+  async getFormatter(name) {
+    const format = this._cliEngine.getFormatter(name)
+
+    return {
+      format(results) {
+        let rulesMeta = null
+
+        results.sort(LinterShell.compareResultsByFilePath)
+
+        return format(results, {
+          get rulesMeta() {
+            if (!rulesMeta) {
+              rulesMeta = createRulesMeta(this._cliEngine.getRules())
+            }
+            return rulesMeta
+          },
+        })
+      },
+    }
+  }
+}
+```
+
+Once we got this change, we can realize the following things:
+
+- (no RFC yet) We can support to print progress state.
+- (no RFC yet) We can support ES modules for custom formatters.
+
+#### ● The other methods
+
+The following methods return `Promise` which gets fulfilled with each result. Once we got this change, we can support ES modules for shareable configs, plugins, and custom parsers without more breaking changes.
+
+- `getConfigForFile()`
+- `isPathIgnored()`
+
+The following methods return a `Promise`. Once we got this change, we can use asynchronous `fs` API to write files.
+
+- `static outputFixes()`
+
+The following methods are as-is.
+
+- `static getErrorResults()`
+
+The following methods are removed because those don't fit the new API.
+
+- `addPlugin()` ... This method has caused to confuse people. We have introduced this method to add plugin implementations and expected people to use this method to test plugins. But people have often thought that this method loads a new plugin for the following linting so they can use plugins rules without `plugins` setting. And this method is only one that mutates the state of `CLIEngine` objects and messes all caches. Therefore, this RFC moves this functionality to a constructor option. See also "[Constructor](#-constructor)" section.
+- `getRules()` ... This method returns the map that contains core rules and the rules of the plugin that the previous `executeOnFiles()` method call used. This behavior is surprised and forces us to store the config objects that the previous `executeOnFiles()` method call used. This proposal removes this method and a separated RFC (maybe [RFC47]) will add the successor.
+- `resolveFileGlobPatterns()` ... ESLint doesn't use this logic since `v6.0.0`, but it has stayed there for backward compatibility. Once [RFC20] is implemented, what ESLint iterates and what the glob of this method iterates will be different, then it will confuse users. This is good timing to remove the legacy.
+
+#### ● New methods
+
+- `static compareResultsByFilePath()` ... This method receives two lint results then returns `+1`, `-1`, or `0`. This method is intended to use in order to sort results.
```

Comment:
I’m not sure I understand the use case for this method. Do you mind expanding or showing an example?

Reply:
My understanding is that this is explained as part of [`executeOnFiles()`](https://github.com/eslint/rfcs/pull/40/files#diff-a206adca92f13d7d6f95fe808b733c13R172):

> ESLint doesn't guarantee the order of the lint results in the array. If we implemented parallel linting, the result of the file that finished linting earlier will be in the front of the array. For backward compatibility, the wrapper of formatters sorts the results then gives the formatters the sorted results. See also [getFormatter()](#-the-getformatter-method). And we provide [compareResultsByFilePath()](#-new-methods) method to sort the results in the same order as `CLIEngine`.

So right now this would effectively be a no-op since `LinterShell#executeOnFiles()` is passes through to `CLIEngine`, but we can add this now for forward-compatibility when results might come in any order.

(Please correct me if I'm wrong @mysticatea)

Reply:
Got it. Thanks!